### ISSUE TITLE
Subqueries in UPDATE/DELETE that touch multiple nodes

### ISSUE THREAD

**Issue by onderkalaci** [citusdata/citus#1947]
With #1127 we've implemented this for queries that hit single node.

We might consider implementing this for queries that hit multiple nodes. Theoretically, with #1804 this might be simpler but needs more thinking for sure.

An example that came on Slack:
```SQL
DELETE FROM reference_table 
WHERE some_id IN (SELECT "id" FROM distributed_table WHERE tenant_id = '6f78ba01-10c3-477f-821e-3321cf572aca');
```

Or a test from the regression test file:
```SQL
UPDATE summary_table SET average_value = average_query.average FROM (
        SELECT avg(value) AS average FROM raw_table) average_query;
ERROR:  subqueries are not supported in modifications across multiple shards
DETAIL:  Consider using an equality filter on partition column "id" to target a single shard.
```

**Comment by marcocitus** [comment]
Seems like a duplicate of https://github.com/citusdata/citus/issues/1740 , though maybe we close #1740 since this contains more details.

**Comment by saicitus** [comment]
@onderkalaci @ozgune @sumedhpathak  This has come up with 4 prospect/customer engagements. Saw this both with UPDATE and DELETE. May be we should edit the title to also capture the DELETE case.

**Comment by onderkalaci** [comment]
Just noting that the following also seems to be solvable in the same way:
```SQL

WITH cte AS (
	SELECT value_2 FROM users_table WHERE user_id IN (1, 2, 3)
)
DELETE FROM modify_table WHERE id IN (SELECT value_2 FROM cte);
ERROR:  common table expressions are not supported in distributed modifications
```

**Comment by marcocitus** [comment]
As of #2006, there is a slightly hacky workaround: wrapping the UPDATE/DELETE and the subquery in CTEs:

```sql
WITH ids_to_delete AS (
  SELECT "id" FROM distributed_table WHERE tenant_id = '6f78ba01-10c3-477f-821e-3321cf572aca'
),
do_delete AS (
  DELETE FROM reference_table WHERE some_id IN (SELECT * FROM ids_to_delete) RETURNING NULL
)
SELECT count(*) FROM do_delete;
```

```sql
WITH average_queryAS (
  SELECT avg(value) AS average FROM raw_table
),
do_update AS (
  UPDATE summary_table SET average_value = average_query.average FROM average_query RETURNING NULL
)
SELECT count(*) FROM do_update;
```

```sql
WITH cte AS (
  SELECT value_2 FROM users_table WHERE user_id IN (1, 2, 3)
),
do_delete AS (
  DELETE FROM modify_table WHERE id IN (SELECT value_2 FROM cte) RETURNING NULL
)
SELECT count(*) FROM do_delete;
```

**Comment by marcocitus** [comment]
Fixed by #2124 and #2137
